pac4j-core/src/main/java/org/pac4j/core/profile/UserProfile.java

```diff
@@ -154,12 +156,18 @@ public String getTypedId() {
     }
 
     /**
-     * Get attributes as immutable map.
+     * Get all attributes as immutable map.
      * 
      * @return the immutable attributes
      */
     public Map<String, Object> getAttributes() {
-        return Collections.unmodifiableMap(this.attributes);
+        final Map<String, Object> newAttributes = new HashMap<>();
+        for (Map.Entry<String, Object> entries : this.attributes.entrySet()) {
+            final String key = entries.getKey();
+            final Object value = getAttribute(key);
+            newAttributes.put(key, value);
+        }
+        return Collections.unmodifiableMap(newAttributes);
```

Comment:
Yes, as we create a new list, adding an attribute won't affect the original attributes map. But I prefer to prohibit direct addition to making obvious that the `addAttribute` method should be used.